File: TreeDistance.cpp
#include<bits/stdc++.h>
using namespace std;
#define fast_io ios_base::sync_with_stdio(false);cin.tie(NULL)

void init() {
fast_io;
#ifndef ONLINE_JUDGE
freopen("input.txt", "r", stdin);
freopen("output.txt", "w", stdout);
#endif // ONLINE_JUDGE
}


pair<int,int>bfs(int node, int n, vector<int>adj[]){
	   queue<pair<int,int>>q;
	   q.push({node,0});

	   vector<bool>vis(n+1, false);

	   vis[node]=true;

	   pair<int,int>u;

	   while(!q.empty()){
	   	     u = q.front();
	   	     q.pop();

	   	     for(auto child: adj[u.first]){
	   	     	   if(!vis[child]){ 
                       q.push({child,u.second+1});
                       vis[child]=true;
	   	     	   }
	   	     }

	   }

	   return u;
}


void dfs(int par, int node, int d, int c, vector<int>adj[], vector<vector<int>>&dp){
	    dp[node][c] = d;
	    for(auto child: adj[node]){
	    	  if(child != par){
	    	  	  dfs(node, child, d+1, c, adj, dp);
	    	  }
	    }
}

int main(){
    init();

    int n;
    cin >> n;

    vector<int>adj[n+1];

    for(int i=1;i<n;i++){
    	  int a,b;
    	  cin >> a >> b;
    	  adj[a].push_back(b);
    	  adj[b].push_back(a);
    }
  
     pair<int,int>p = bfs(1,n,adj);

     pair<int,int>pp = bfs(p.first, n, adj);
   


     vector<vector<int>>dp(n+1, vector<int>(2,0));

     dfs(0,p.first, 0, 0, adj, dp);

     dfs(0,pp.first,0,1,adj,dp);

     for(int i=1;i<=n;i++)
     	   cout << max(dp[i][0],dp[i][1]) << " ";

    return 0;
}

File: TreeDistance.py
import queue

def bfs(node, n, adj):
    q = queue.Queue()
    q.put((node, 0))
    
    vis = [False] * (n + 1)
    vis[node] = True
    
    u = (0, 0)
    
    while not q.empty():
        u = q.get()
        
        for child in adj[u[0]]:
            if not vis[child]:
                q.put((child, u[1] + 1))
                vis[child] = True
                
    return u

def dfs(par, node, d, c, adj, dp):
    dp[node][c] = d
    
    for child in adj[node]:
        if child != par:
            dfs(node, child, d + 1, c, adj, dp)

def main():
    n = int(input())
    
    adj = [[] for _ in range(n + 1)]
    
    for i in range(1, n):
        a, b = map(int, input().split())
        adj[a].append(b)
        adj[b].append(a)
    
    p = bfs(1, n, adj)
    
    pp = bfs(p[0], n, adj)
    
    dp = [[0] * 2 for _ in range(n + 1)]
    
    dfs(0, p[0], 0, 0, adj, dp)
    
    dfs(0, pp[0], 0, 1, adj, dp)
    
    for i in range(1, n + 1):
        print(max(dp[i][0], dp[i][1]), end=" ")

if __name__ == "__main__":
    main()
